File: Algorithms/Data Structures/trie.go
package main

import "fmt"

/*
	Prefix tree or digital trees or mostly called trie
	It is a type of k-ary search tree.
	There are other forms of this data structure :
	Bitwise tries
	Compressed tries (used in web search engines for indexing.)
	Here I am using an array of 26 characters, assuming only lower case alphabets will be stored.
	If want to store unicode characters use map[byte]*Trie
*/

type Trie struct {
	children [26]*Trie
	word     string
}

func New() *Trie { return &Trie{} }

func (this *Trie) Insert(word string) {
	root := this
	for i := 0; i < len(word); i++ {
		ch := int(word[i] - 'a')
		if root.children[ch] == nil {
			root.children[ch] = New()
		}
		root = root.children[ch]
	}
	root.word = word
}

func (this *Trie) find(word string) bool {
	root := this
	for i := 0; i < len(word); i++ {
		ch := int(word[i] - 'a')
		if root.children[ch] == nil {
			return false
		}
		root = root.children[ch]
	}
	return true
}

func main() {
	trie := New()
	trie.Insert("ball")
	trie.Insert("balloon")
	trie.Insert("zebra")
	fmt.Println("Search ball", trie.find("ball"))
	fmt.Println("Search zebra", trie.find("zebra"))
	fmt.Println("Search balloonm", trie.find("balloon"))
	fmt.Println("Search bat", trie.find("bat"))
}

File: LeetCode/Medium/Implement Trie/c++/trie.cc
class TrieNode {
public:
    bool is_word;
    unordered_map<char, TrieNode*> children;
    TrieNode() : is_word(false) {} 
};

class Trie {
public:
    TrieNode* root;
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* temp = this->root;
        for (char c : word) {
            if (temp->children.find(c) == temp->children.end()) {
                temp->children[c] = new TrieNode();
            }
            temp = temp->children[c];
        }
        temp->is_word = true;
    }
    
    bool search(string word) {
        TrieNode* temp = this->root;
        for (char c : word) {
            if (temp->children.find(c) == temp->children.end()) return false;
            temp = temp->children[c];
        }
        return temp->is_word;
    }
    
    bool startsWith(string prefix) {
        TrieNode* temp = root;
        for (char c : prefix) {
            if (temp->children.find(c) == temp->children.end()) return false;
            temp = temp->children[c];
        }
        return temp->children.size() > 0 || temp->is_word;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
